<center>
<img src="../../img/ods_stickers.jpg">
## Открытый курс по машинному обучению. Сессия № 2
</center>
Автор материала: программист-исследователь Mail.ru Group, старший преподаватель Факультета Компьютерных Наук ВШЭ Юрий Кашницкий. Материал распространяется на условиях лицензии [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Можно использовать в любых целях (редактировать, поправлять и брать за основу), кроме коммерческих, но с обязательным упоминанием автора материала.

# <center>Тема 1. Первичный анализ данных с Pandas</center>
## <center>Практическое задание. Анализ данных пассажиров "Титаника"</center>

**Заполните код в клетках (где написано "Ваш код здесь") и ответьте на вопросы в [веб-форме](https://goo.gl/SbRcpW).**

In [21]:
import numpy as np
import pandas as pd
%matplotlib inline
%config IPCompleter.greedy=True


**Считаем данные из файла в память в виде объекта Pandas.DataFrame**

In [6]:
data = pd.read_csv('../../data/titanic_train.csv',
                  index_col='PassengerId')

**Данные представлены в виде таблицы. Посмотрим на первые 5 строк:**

In [106]:
data.head(6)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   
6                0      0            330877   8.4583   NaN        Q   

             Age_category  AgeCat  
PassengerId                        
1                       1       1  
2                       2       2  
3                       1       1  
4                       2       2  
5                       2       2  
6                       3       2

In [8]:
data.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

**Для примера отберем пассажиров, которые сели в Cherbourg (Embarked=C) и заплатили более 200 у.е. за билет (fare > 200).**

Убедитесь, что Вы понимаете, как эта конструкция работает. <br>
Если нет – посмотрите, как вычисляется выражение в квадратных в скобках.

In [107]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)]

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   
381                 1       1   
558                 0       1   
680                 1       1   
701                 1       1   
717                 1       1   
738                 1       1   
743                 1       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
119                                   Baxter, Mr. Quigg Edmond    male  24.0   
259                                           Ward, Miss. Anna  female  35.0   
300            Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                                 Ryerson, Miss. Emily Borie  female  18.0   
378                                  Widener, Mr. Harry Elkins    male  27.0   
381                                      Bidois, Miss. Rosalie  female  42.0   
558                                        Robbins, Mr. Victor    male   NaN   
680                         Cardeza, Mr. Thomas Drake Martinez    male  36.0   
701          Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female  18.0   
717                              Endres, Miss. Caroline Louise  female  38.0   
738                                     Lesurer, Mr. Gustave J    male  35.0   
743                      Ryerson, Miss. Susan Parker "Suzette"  female  21.0   

             SibSp  Parch    Ticket      Fare            Cabin Embarked  \
PassengerId                                                               
119              0      1  PC 17558  247.5208          B58 B60        C   
259              0      0  PC 17755  512.3292              NaN        C   
300              0      1  PC 17558  247.5208          B58 B60        C   
312              2      2  PC 17608  262.3750  B57 B59 B63 B66        C   
378              0      2    113503  211.5000              C82        C   
381              0      0  PC 17757  227.5250              NaN        C   
558              0      0  PC 17757  227.5250              NaN        C   
680              0      1  PC 17755  512.3292      B51 B53 B55        C   
701              1      0  PC 17757  227.5250          C62 C64        C   
717              0      0  PC 17757  227.5250              C45        C   
738              0      0  PC 17755  512.3292             B101        C   
743              2      2  PC 17608  262.3750  B57 B59 B63 B66        C   

             Age_category  AgeCat  
PassengerId                        
119                     1       1  
259                     2       2  
300                     2       2  
312                     1       1  
378                     1       1  
381                     2       2  
558                     3       2  
680                     2       2  
701                     1       1  
717                     2       2  
738                     2       2  
743                     1       1

**Можно отсортировать этих людей по убыванию платы за билет.**

In [10]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
PassengerId                                                                    
259          35.0      0      0  PC 17755  512.3292              NaN        C  
680          36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.3292             B101        C  
312          18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C

**Пример создания признака.**

In [11]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    else:
        return 3

In [12]:
age_categories = [age_category(age) for age in data.Age]

In [13]:
data['Age_category'] = age_categories

**Другой способ – через `apply`.**

In [15]:
data['Age_category'] = data['Age'].apply(age_category)

**1. Сколько мужчин / женщин находилось на борту?**
- 412 мужчин и 479 женщин
- 314 мужчин и 577 женщин
- 479 мужчин и 412 женщин
- 577 мужчин и 314 женщин

In [27]:
size = data.groupby('Sex').size()
size

Sex
female    314
male      577
dtype: int64

**2. Выведите распределение переменной `Pclass` (социально-экономический статус) и это же распределение, только для мужчин / женщин по отдельности. Сколько было мужчин 2-го класса?**
- 104
- 108 +
- 112
- 125

In [35]:
# Ваш код здесь
df = data[(data['Sex']=='male') & (data['Pclass']==2)]


In [37]:
df.shape

(108, 12)

**3. Каковы медиана и стандартное отклонение платежей (`Fare`)? Округлите до 2 десятичных знаков.**
- Медиана – 14.45, стандартное отклонение – 49.69 ???
- Медиана – 15.1, стандартное отклонение – 12.15
- Медиана – 13.15, стандартное отклонение – 35.3
- Медиана – 17.43, стандартное отклонение – 39.1

In [50]:
# Ваш код здесь
descFare = data['Fare'].describe()
[descFare.median(), descFare.std()]
descFare


count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

In [45]:
descFare

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

**4. Правда ли, что люди моложе 30 лет выживали чаще, чем люди старше 60 лет? Каковы доли выживших в обеих группах?**
- 22.7% среди молодых и 40.6% среди старых
- 40.6% среди молодых и 22.7% среди старых +
- 35.3% среди молодых и 27.4% среди старых
- 27.4% среди молодых и 35.3% среди старых

In [108]:
foo = lambda age:  1 if age<30 else (3 if age>60 else 2)
data['AgeCat'] = data['Age'].apply(foo)
#data.groupby(['AgeCat', 'Survived']).describe()
count = data.shape[0]
cstb = pd.crosstab(data['AgeCat'], data['Survived'])
cstb.apply(lambda r: 100*r/r.sum(), axis=1)

Survived          0          1
AgeCat                        
1         59.375000  40.625000
2         62.680412  37.319588
3         77.272727  22.727273

**5. Правда ли, что женщины выживали чаще мужчин? Каковы доли выживших в обеих группах?**
- 30.2% среди мужчин и 46.2% среди женщин
- 35.7% среди мужчин и 74.2% среди женщин
- 21.1% среди мужчин и 46.2% среди женщин
- 18.9% среди мужчин и 74.2% среди женщин +

In [109]:
cstb = pd.crosstab(data['Sex'], data['Survived'].apply(lambda d: 100* d/d.sum(), axis=1)
cstb

Survived    0    1
Sex               
female     81  233
male      468  109

**6. Найдите самое популярное имя среди пассажиров Титаника мужского пола?**
- Charles
- Thomas
- William
- John

In [105]:
# Ваш код здесь
data['Name'].value_counts()
data['Name'].str.

PassengerId
1                           [Braund,, Mr., Owen, Harris]
2      [Cumings,, Mrs., John, Bradley, (Florence, Bri...
3                             [Heikkinen,, Miss., Laina]
4      [Futrelle,, Mrs., Jacques, Heath, (Lily, May, ...
5                          [Allen,, Mr., William, Henry]
6                                   [Moran,, Mr., James]
7                           [McCarthy,, Mr., Timothy, J]
8                    [Palsson,, Master., Gosta, Leonard]
9      [Johnson,, Mrs., Oscar, W, (Elisabeth, Vilhelm...
10             [Nasser,, Mrs., Nicholas, (Adele, Achem)]
11                  [Sandstrom,, Miss., Marguerite, Rut]
12                          [Bonnell,, Miss., Elizabeth]
13                   [Saundercock,, Mr., William, Henry]
14                      [Andersson,, Mr., Anders, Johan]
15            [Vestrom,, Miss., Hulda, Amanda, Adolfina]
16                 [Hewlett,, Mrs., (Mary, D, Kingcome)]
17                              [Rice,, Master., Eugene]
18                 

**7. Сравните графически распределение стоимости билетов и возраста у спасенных и у погибших. Средний возраст погибших выше, верно?**
- Да 
- Нет

In [17]:
# Ваш код здесь

**8. Как отличается средний возраст мужчин / женщин в зависимости от класса обслуживания? Выберите верные утверждения:**
- В среднем мужчины 1-го класса старше 40 лет
- В среднем женщины 1-го класса старше 40 лет
- Мужчины всех классов в среднем старше женщин того же класса
- В среднем люди в 1 классе старше, чем во 2-ом, а те старше представителей 3-го класса

In [18]:
# Ваш код здесь